pkg/storage/utils/volumes.go

```diff
@@ -74,58 +55,137 @@ func GetVolumes(obj interface{}, client kubecli.KubevirtClient, opts ...VolumeOp
-	if needsBackendPVC(vm.Spec.Template.Spec, opts) {
-		if client == nil {
-			return []v1.Volume{}, fmt.Errorf("no client provided")
+	return getVolumes(vm, vm.Spec.Template.Spec, client, opts...)
+}
+
+// GetSnapshotVirtualMachineVolumes returns all volumes of a Snapshot VM except the special ones based on volume options
+func GetSnapshotVirtualMachineVolumes(vm *snapshotv1.VirtualMachine, client kubecli.KubevirtClient, opts ...VolumeOption) ([]v1.Volume, error) {
+	return getVolumes(vm, vm.Spec.Template.Spec, client, opts...)
+}
+
+// GetVirtualMachineInstanceVolumes returns all volumes of a VMI except the special ones based on volume options
+func GetVirtualMachineInstanceVolumes(vmi *v1.VirtualMachineInstance, opts ...VolumeOption) ([]v1.Volume, error) {
+	return getVolumes(vmi, vmi.Spec, nil, opts...)
+}
+
+func getVolumes(obj metav1.Object, vmiSpec v1.VirtualMachineInstanceSpec, client kubecli.KubevirtClient, opts ...VolumeOption) ([]v1.Volume, error) {
+	var enumeratedVolumes []v1.Volume
+
+	if needsRegularVolumes(vmiSpec, opts) {
+		for _, volume := range vmiSpec.Volumes {
+			enumeratedVolumes = append(enumeratedVolumes, volume)
 		}
-		vmi, err = client.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+	}
+
+	if needsBackendPVC(vmiSpec, opts) {
+		backendVolumeName, err := getBackendPVCName(obj, client)
 		if err != nil {
-			return []v1.Volume{}, err
+			return enumeratedVolumes, err
+		}
+		if backendVolumeName != "" {
+			enumeratedVolumes = append(enumeratedVolumes, *createPVCVolume(backendVolumeName))
 		}
 	}
-	return GetVirtualMachineInstanceVolumes(vmi, opts...), nil
+
+	return enumeratedVolumes, nil
 }
 
-// GetSnapshotVirtualMachineVolumes returns all volumes of a Snapshot VM except the special ones based on volume options
-func GetSnapshotVirtualMachineVolumes(vm *snapshotv1.VirtualMachine, client kubecli.KubevirtClient, opts ...VolumeOption) ([]v1.Volume, error) {
-	var err error
-	vmi := &v1.VirtualMachineInstance{ObjectMeta: metav1.ObjectMeta{Name: vm.Name}, Spec: vm.Spec.Template.Spec}
-	if needsBackendPVC(vm.Spec.Template.Spec, opts) {
+func getBackendPVCName(obj metav1.Object, client kubecli.KubevirtClient) (string, error) {
+	switch obj := obj.(type) {
+	case *v1.VirtualMachineInstance:
+		return backendstorage.CurrentPVCName(obj), nil
+	default:
+		// TODO: This could be way more simpler if the backend PVC name was accessible from the VM spec/status.
+		// Refactor this once the backend PVC is more accessible.
 		if client == nil {
-			return []v1.Volume{}, fmt.Errorf("no client provided")
+			return "", fmt.Errorf("no client provided")
 		}
-		vmi, err = client.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+		pvcs, err := client.CoreV1().PersistentVolumeClaims(obj.GetNamespace()).List(context.Background(), metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", backendstorage.PVCPrefix, obj.GetName()),
```

Comment:
Labels have character limits in both key and value but here I'm just setting the label selector. I don't think we should worry about char limits here.

Reply:
No I do think you need to worry about it here. Since if the `obj.GetName()` + prefix is > 63 characters, then it will never match. Now obviously the code that generates the label in the first place will also fail, and should get fixed. And that same fix needs to be applied here.

I recently fixed a label length related issue in the export code, I don't want to re-introduce an issue with this.

Reply:
Individual key and values in labels are limited to 63 chars but not the label selector. That would make it impossible to filter resources by labels whose key+value surpasses the 63-char limit. Here, the label key is always `persistent-state-for`, while the value is the VM name.

Reply:
Okay, discussed this offline with @awels and he is right. It's true that since the VM name can be larger than 63 chars this label can fail. Since this is an issue in all backend-storage logic I hope we can leave it for a follow-up PR.